{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://github.com/phbradley/TCRdock/blob/main/datasets_from_the_paper/table_S1_structure_benchmark_complexes.csv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 0. Setup: helper methods\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "import polars as pl\n",
    "from tcr_format_parsers.common.MHCCodeConverter import (\n",
    "    HLASequenceDBConverter,\n",
    "    H2SequenceDictConverter,\n",
    ")\n",
    "from tcr_format_parsers.common.TCRUtils import standardize_tcr\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "import MDAnalysis as mda\n",
    "\n",
    "def format_pdb_df(df):\n",
    "    df = df.with_columns(\n",
    "        pl.when(pl.col(\"mhc_type\") == \"MH1\")\n",
    "        .then(pl.lit(\"I\"))\n",
    "        .when(pl.col(\"mhc_type\") == \"MH2\")\n",
    "        .then(pl.lit(\"II\"))\n",
    "        .otherwise(None)\n",
    "        .alias(\"mhc_class\"),\n",
    "    )\n",
    "\n",
    "    # df = df.filter(\n",
    "    #     (pl.col(\"mhc_chain1\").is_not_null())\n",
    "    #     & (pl.col(\"mhc_chain2\").is_not_null())\n",
    "    # )\n",
    "\n",
    "    df = df.group_by(\"pdb\").agg(\n",
    "        pl.col(\"Bchain\").drop_nulls().first(),\n",
    "        pl.col(\"Achain\").drop_nulls().first(),\n",
    "        pl.col(\"mhc_chain1\").drop_nulls().first(),\n",
    "        pl.col(\"mhc_chain2\").drop_nulls().first(),\n",
    "        pl.col(\"antigen_chain\").drop_nulls().first(),\n",
    "        pl.col(\"mhc_class\").drop_nulls().first(),\n",
    "        pl.col(\"mhc_chain1_organism\").drop_nulls().first().alias(\"mhc_1_species\"),\n",
    "        pl.col(\"mhc_chain2_organism\").drop_nulls().first().alias(\"mhc_2_species\"),\n",
    "        pl.col(\"alpha_organism\").drop_nulls().first().alias(\"tcr_1_species\"),\n",
    "        pl.col(\"beta_organism\").drop_nulls().first().alias(\"tcr_2_species\"),\n",
    "    )\n",
    "\n",
    "    df = df.with_columns(\n",
    "        pl.when(pl.col(\"mhc_1_species\") == \"homo sapiens\")\n",
    "        .then(pl.lit(\"human\"))\n",
    "        .when(pl.col(\"mhc_1_species\") == \"mus musculus\")\n",
    "        .then(pl.lit(\"mouse\"))\n",
    "        .otherwise(None)\n",
    "        .alias(\"mhc_1_species\"),\n",
    "        pl.when(pl.col(\"mhc_2_species\") == \"homo sapiens\")\n",
    "        .then(pl.lit(\"human\"))\n",
    "        .when(pl.col(\"mhc_2_species\") == \"mus musculus\")\n",
    "        .then(pl.lit(\"mouse\"))\n",
    "        .otherwise(None)\n",
    "        .alias(\"mhc_2_species\"),\n",
    "        pl.when(pl.col(\"tcr_1_species\") == \"homo sapiens\")\n",
    "        .then(pl.lit(\"human\"))\n",
    "        .when(pl.col(\"tcr_1_species\") == \"mus musculus\")\n",
    "        .then(pl.lit(\"mouse\"))\n",
    "        .otherwise(None)\n",
    "        .alias(\"tcr_1_species\"),\n",
    "        pl.when(pl.col(\"tcr_2_species\") == \"homo sapiens\")\n",
    "        .then(pl.lit(\"human\"))\n",
    "        .when(pl.col(\"tcr_2_species\") == \"mus musculus\")\n",
    "        .then(pl.lit(\"mouse\"))\n",
    "        .otherwise(None)\n",
    "        .alias(\"tcr_2_species\"),\n",
    "    )\n",
    "\n",
    "    df = df.with_columns(\n",
    "        pl.when(pl.col(\"mhc_class\") == \"II\")\n",
    "        .then(pl.lit(\"alpha\"))\n",
    "        .otherwise(pl.lit(\"heavy\"))\n",
    "        .alias(\"mhc_1_chain\"),\n",
    "        pl.when(pl.col(\"mhc_class\") == \"II\")\n",
    "        .then(pl.lit(\"beta\"))\n",
    "        .otherwise(pl.lit(\"light\"))\n",
    "        .alias(\"mhc_2_chain\"),\n",
    "        pl.lit(True).alias(\"cognate\"),\n",
    "        pl.lit(\"alpha\").alias(\"tcr_1_chain\"),\n",
    "        pl.lit(\"beta\").alias(\"tcr_2_chain\"),\n",
    "    )\n",
    "\n",
    "    df = df.with_columns(\n",
    "        pl.col(\"antigen_chain\")\n",
    "        .str.split(\"|\")\n",
    "        .list.first()\n",
    "        .str.strip_chars()\n",
    "        .alias(\"antigen_chain\")\n",
    "    )\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def get_pdb_date(row):\n",
    "    r = requests.get(\"https://data.rcsb.org/rest/v1/core/entry/\" + row[\"pdb\"])\n",
    "    r.raise_for_status()\n",
    "    new_row = row.copy()\n",
    "    new_row[\"pdb_date\"] = r.json()[\"rcsb_accession_info\"][\n",
    "        \"initial_release_date\"\n",
    "    ]\n",
    "\n",
    "    return pl.DataFrame(new_row).with_columns(\n",
    "        pl.col(\"pdb_date\").str.to_datetime().alias(\"pdb_date\")\n",
    "    )\n",
    "\n",
    "\n",
    "def parse_chain(chain):\n",
    "    if \"[\" in chain:\n",
    "        \n",
    "        return chain.split(\"[auth \")[1][0]\n",
    "        # if can have multi-letter chains\n",
    "        # return chain.split(\"[auth \")[1].split(\"]\")[0]\n",
    "    else:\n",
    "        return chain.replace(\" \", \"\")\n",
    "\n",
    "\n",
    "def parse_fasta_description(description):\n",
    "    chain_token = description.split(\"|\")[1]\n",
    "\n",
    "    if chain_token.startswith(\"Chain \"):\n",
    "        return list(parse_chain(chain_token.split(\"Chain \")[1]))\n",
    "    else:\n",
    "        chains = chain_token.split(\"Chains \")[1].split(\",\")\n",
    "        chain_list = [parse_chain(chain) for chain in chains]\n",
    "\n",
    "        return chain_list\n",
    "\n",
    "\n",
    "def get_fasta_seq(\n",
    "    pdb_id,\n",
    "    antigen_chain_id,\n",
    "    mhc_chain1_id,\n",
    "    mhc_chain2_id,\n",
    "    Achain_id,\n",
    "    Bchain_id,\n",
    "):\n",
    "    r = requests.get(\"https://www.rcsb.org/fasta/entry/\" + pdb_id)\n",
    "\n",
    "    r.raise_for_status()\n",
    "\n",
    "    fasta_sequences = SeqIO.parse(StringIO(r.text), \"fasta\")\n",
    "\n",
    "    seq_dict = {}\n",
    "    for fasta in fasta_sequences:\n",
    "        chains = parse_fasta_description(fasta.description)\n",
    "        for chain in chains:\n",
    "            seq_dict[chain] = str(fasta.seq)\n",
    "\n",
    "    return {\n",
    "        \"peptide_seq\": seq_dict[antigen_chain_id] if antigen_chain_id is not None else \"\",\n",
    "        \"mhc_1_seq\": seq_dict[mhc_chain1_id] if mhc_chain1_id is not None else \"\",\n",
    "        \"mhc_2_seq\": seq_dict[mhc_chain2_id] if mhc_chain2_id is not None else \"\",\n",
    "        \"tcr_1_seq\": seq_dict[Achain_id] if Achain_id is not None else \"\",\n",
    "        \"tcr_2_seq\": seq_dict[Bchain_id] if Bchain_id is not None else \"\",\n",
    "    }\n",
    "\n",
    "\n",
    "SEQ_STRUCT = pl.Struct(\n",
    "    {\n",
    "        \"peptide_seq\": pl.String,\n",
    "        \"mhc_1_seq\": pl.String,\n",
    "        \"mhc_2_seq\": pl.String,\n",
    "        \"tcr_1_seq\": pl.String,\n",
    "        \"tcr_2_seq\": pl.String,\n",
    "    }\n",
    ")\n",
    "\n",
    "def format_seqs(df, skip_peptide=False):\n",
    "    df = df.with_columns(\n",
    "        pl.struct(\n",
    "            pl.col(\"pdb\"),\n",
    "            pl.col(\"Bchain\"),\n",
    "            pl.col(\"Achain\"),\n",
    "            pl.col(\"antigen_chain\"),\n",
    "            pl.col(\"mhc_chain1\"),\n",
    "            pl.col(\"mhc_chain2\"),\n",
    "        )\n",
    "        .map_elements(\n",
    "            lambda x: get_fasta_seq(\n",
    "                x[\"pdb\"],\n",
    "                x[\"antigen_chain\"],\n",
    "                x[\"mhc_chain1\"],\n",
    "                x[\"mhc_chain2\"],\n",
    "                x[\"Achain\"],\n",
    "                x[\"Bchain\"],\n",
    "            ),\n",
    "            return_dtype=SEQ_STRUCT,\n",
    "            skip_nulls=False,\n",
    "        )\n",
    "        .alias(\"chain_seqs\"),\n",
    "    ).unnest(\"chain_seqs\").with_columns(\n",
    "        pl.when(pl.col(\"peptide_seq\") == \"\")\n",
    "        .then(pl.lit(None))\n",
    "        .otherwise(pl.col(\"peptide_seq\"))\n",
    "        .alias(\"peptide_seq\"),\n",
    "        pl.when(pl.col(\"mhc_1_seq\") == \"\")\n",
    "        .then(pl.lit(None))\n",
    "        .otherwise(pl.col(\"mhc_1_seq\"))\n",
    "        .alias(\"mhc_1_seq\"),\n",
    "        pl.when(pl.col(\"mhc_2_seq\") == \"\")\n",
    "        .then(pl.lit(None))\n",
    "        .otherwise(pl.col(\"mhc_2_seq\"))\n",
    "        .alias(\"mhc_2_seq\"),\n",
    "        pl.when(pl.col(\"tcr_1_seq\") == \"\")\n",
    "        .then(pl.lit(None))\n",
    "        .otherwise(pl.col(\"tcr_1_seq\"))\n",
    "        .alias(\"tcr_1_seq\"),\n",
    "        pl.when(pl.col(\"tcr_2_seq\") == \"\")\n",
    "        .then(pl.lit(None))\n",
    "        .otherwise(pl.col(\"tcr_2_seq\"))\n",
    "        .alias(\"tcr_2_seq\"),\n",
    "    )\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def remove_peptide_from_chains(row):\n",
    "    new_row = row.copy()\n",
    "\n",
    "    if row[\"mhc_1_seq\"] is not None and row[\"peptide\"] in row[\"mhc_1_seq\"]:\n",
    "        warnings.warn(f\"Peptide found in MHC 1 sequence for PDB {row['pdb']} at position {row['mhc_1_seq'].index(row['peptide'])}\")\n",
    "        index_of_peptide = row[\"mhc_1_seq\"].index(row[\"peptide\"])\n",
    "        new_row[\"mhc_1_seq\"] = new_row[\"mhc_1_seq\"][\n",
    "            index_of_peptide + len(row[\"peptide\"]) :\n",
    "        ]\n",
    "    if row[\"mhc_2_seq\"] is not None and row[\"peptide\"] in row[\"mhc_2_seq\"]:\n",
    "        warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
    "        index_of_peptide = row[\"mhc_2_seq\"].index(row[\"peptide\"])\n",
    "        new_row[\"mhc_2_seq\"] = new_row[\"mhc_2_seq\"][\n",
    "            index_of_peptide + len(row[\"peptide\"]) :\n",
    "        ]\n",
    "    if row[\"tcr_1_seq\"] is not None and row[\"peptide\"] in row[\"tcr_1_seq\"]:\n",
    "        warnings.warn(f\"Peptide found in TCR 1 sequence for PDB {row['pdb']} at position {row['tcr_1_seq'].index(row['peptide'])}\")\n",
    "        index_of_peptide = row[\"tcr_1_seq\"].index(row[\"peptide\"])\n",
    "        new_row[\"tcr_1_seq\"] = new_row[\"tcr_1_seq\"][\n",
    "            index_of_peptide + len(row[\"peptide\"]) :\n",
    "        ]\n",
    "    if row[\"tcr_2_seq\"] is not None and row[\"peptide\"] in row[\"tcr_2_seq\"]:\n",
    "        warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
    "        index_of_peptide = row[\"tcr_2_seq\"].index(row[\"peptide\"])\n",
    "        new_row[\"tcr_2_seq\"] = new_row[\"tcr_2_seq\"][\n",
    "            index_of_peptide + len(row[\"peptide\"]) :\n",
    "        ]\n",
    "    return pl.DataFrame(new_row)\n",
    "\n",
    "\n",
    "def infer_correct_mhc(row, human_conv, mouse_conv):\n",
    "    mhc1 = row[\"mhc_1_seq\"]\n",
    "    mhc2 = row[\"mhc_2_seq\"]\n",
    "\n",
    "    if row[\"organism\"] == \"human\":\n",
    "        mhc_1_inf = human_conv.get_mhc_allele(\n",
    "            mhc1, chain=row[\"mhc_1_chain\"], top_only=True\n",
    "        )\n",
    "    else:\n",
    "        mhc_1_inf = mouse_conv.get_mhc_allele(\n",
    "            mhc1, chain=row[\"mhc_1_chain\"], top_only=True\n",
    "        )\n",
    "\n",
    "    if row[\"organism\"] == \"human\":\n",
    "        if row[\"mhc_class\"] == \"I\" and row['mhc_2_seq'] is None:\n",
    "            mhc_2_inf = {\n",
    "                \"mhc_2_match_seq\": None,\n",
    "                \"mhc_2_name\" : None,\n",
    "                \"mhc_2_match_size\": None,\n",
    "                \"mhc_2_match_proportion\": None,\n",
    "                \"mhc_2_status\": None,\n",
    "                \"mhc_2_maxres\": None,\n",
    "            }\n",
    "        else:\n",
    "            mhc_2_inf = human_conv.get_mhc_allele(\n",
    "                mhc2, chain=row[\"mhc_2_chain\"], top_only=True\n",
    "            )\n",
    "    else:\n",
    "        mhc_2_inf = mouse_conv.get_mhc_allele(\n",
    "            mhc2, chain=row[\"mhc_2_chain\"], top_only=True\n",
    "        )\n",
    "\n",
    "    new_row = row.copy()\n",
    "\n",
    "    new_row[\"mhc_1_match_seq\"] = mhc_1_inf[\"seq\"]\n",
    "    new_row[\"mhc_1_name\"] = mhc_1_inf[\"name\"]\n",
    "    new_row[\"mhc_1_match_size\"] = mhc_1_inf[\"match_size\"]\n",
    "    new_row[\"mhc_1_match_proportion\"] = (\n",
    "        (mhc_1_inf[\"match_size\"] / len(mhc1))\n",
    "        if mhc_1_inf[\"match_size\"] is not None\n",
    "        else None\n",
    "    )\n",
    "    new_row[\"mhc_1_status\"] = mhc_1_inf[\"sequence_status\"]\n",
    "    new_row[\"mhc_1_name_maxres\"] = mhc_1_inf[\"max_resolution_name\"]\n",
    "\n",
    "    new_row[\"mhc_2_match_seq\"] = mhc_2_inf[\"seq\"]\n",
    "    new_row[\"mhc_2_name\"] = mhc_2_inf[\"name\"]\n",
    "    new_row[\"mhc_2_match_size\"] = mhc_2_inf[\"match_size\"]\n",
    "    new_row[\"mhc_2_match_proportion\"] = (\n",
    "        (mhc_2_inf[\"match_size\"] / len(mhc2))\n",
    "        if mhc_2_inf[\"match_size\"] is not None\n",
    "        else None\n",
    "    )\n",
    "    new_row[\"mhc_2_status\"] = mhc_2_inf[\"sequence_status\"]\n",
    "    new_row[\"mhc_2_maxres\"] = mhc_2_inf[\"max_resolution_name\"]\n",
    "    return pl.DataFrame(new_row)\n",
    "\n",
    "\n",
    "def download_pdb(row, path):\n",
    "\n",
    "    r = requests.get(f\"https://files.rcsb.org/download/{row[\"pdb\"]}.pdb\")\n",
    "    suffix = \".pdb\"\n",
    "    try:\n",
    "        r.raise_for_status()\n",
    "    except Exception as e:\n",
    "        r = requests.get(f\"https://files.rcsb.org/download/{row[\"pdb\"]}.cif\")\n",
    "        suffix = \".cif\"\n",
    "        r.raise_for_status()\n",
    "    with open(path / (row[\"pdb\"] + suffix), \"wb\") as f:\n",
    "        f.write(r.content)\n",
    "    return pl.DataFrame(row)\n",
    "\n",
    "def get_true_mda_universe(pdb_id, root_path):\n",
    "    # Favor PDB since it doesn't have multiple residue with same ID issue\n",
    "    if (root_path / (pdb_id + \".pdb\")).exists():\n",
    "        suffix = \".pdb\"\n",
    "\n",
    "    else:\n",
    "        suffix = \".cif\"\n",
    "\n",
    "    return mda.Universe((root_path / (pdb_id + suffix)).as_posix())\n",
    "\n",
    "def update_df_from_k_v(\n",
    "    df,\n",
    "    primary_key_colname,\n",
    "    primary_key,\n",
    "    k,\n",
    "    v,\n",
    "):\n",
    "    df = pl.concat([df.filter(pl.col(primary_key_colname) == primary_key).with_columns(\n",
    "        pl.lit(v).alias(k)\n",
    "    ), df.filter(pl.col(primary_key_colname) != primary_key)], how=\"vertical_relaxed\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Import data, query IEDB locally\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 634/634 [00:33<00:00, 18.88it/s]\n"
     ]
    }
   ],
   "source": [
    "import polars as pl\n",
    "from tcr_format_parsers.common.TriadUtils import FORMAT_ANTIGEN_COLS\n",
    "from tcr_format_parsers.common.TCRUtils import hash_tcr_sequence\n",
    "import polars as pl\n",
    "from pathlib import Path\n",
    "from datetime import datetime, timezone\n",
    "from mdaf3.FeatureExtraction import serial_apply\n",
    "\n",
    "IMGT_HLA_PATH = Path(\"/tgen_labs/altin/alphafold3/IMGTHLA\")\n",
    "\n",
    "\n",
    "\n",
    "schema_overrides = {\n",
    "    \"Gchain\": pl.String,\n",
    "    \"Dchain\": pl.String,\n",
    "}\n",
    "null_values = [\"NA\", \"unknown\", \"NOT\"]\n",
    "\n",
    "all_pdb = format_pdb_df(\n",
    "    pl.read_csv(\n",
    "        \"raw/db_summary.dat\",\n",
    "        schema_overrides=schema_overrides,\n",
    "        null_values=null_values,\n",
    "        separator=\"\\t\",\n",
    "    )\n",
    ")\n",
    "\n",
    "all_pdb = serial_apply(\n",
    "    all_pdb,\n",
    "    get_pdb_date,\n",
    ")\n",
    "\n",
    "phil_pdb = (\n",
    "    pl.read_csv(\n",
    "        \"raw/table_S1_structure_benchmark_complexes.csv\",\n",
    "    )\n",
    "    .rename({\"pdbid\": \"pdb\"})\n",
    "    .with_columns(\n",
    "        pl.when(pl.col(\"mhc_class\") == 1)\n",
    "        .then(pl.lit(\"I\"))\n",
    "        .otherwise(pl.lit(\"II\"))\n",
    "        .alias(\"mhc_class\"),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Recreating Phil Bradley's Elife dataset\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rows STCRDab found\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stcr_dat_phil = phil_pdb.join(all_pdb, on=\"pdb\", how=\"inner\")\n",
    "\n",
    "mhc_1_missing_b2m = [\n",
    "    \"4n0c\",\n",
    "    \"4mvb\",\n",
    "    \"3tpu\",\n",
    "    \"3tf7\",\n",
    "    \"4n5e\",\n",
    "    \"4mxq\",\n",
    "    \"3tfk\",\n",
    "    \"2oi9\",\n",
    "]\n",
    "\n",
    "pre_fasta_corrections = {\n",
    "    \"3tf7\": {\"mhc_chain1\": \"E\", \"Achain\": None, \"Bchain\": None},\n",
    "    # antigen chain located on MHC 2 / TCR 2 chain\n",
    "    # when we later look for the peptide we can find it by aligning to this chain\n",
    "    \"6bga\": {\"antigen_chain\": \"B\"},\n",
    "    \"3pl6\": {\"antigen_chain\": \"D\"},\n",
    "    \"3o6f\": {\"antigen_chain\": \"B\"},\n",
    "    \"6dfw\": {\"antigen_chain\": \"D\"},\n",
    "    \"3c5z\": {\"antigen_chain\": \"D\"},\n",
    "    \"3rdt\": {\"antigen_chain\": \"D\"},\n",
    "    \"6dfx\": {\"antigen_chain\": \"E\"},\n",
    "    \"3c60\": {\"antigen_chain\": \"D\"},\n",
    "    \"4grl\": {\"antigen_chain\": \"D\"},\n",
    "    \"6mnn\": {\"antigen_chain\": \"D\"},\n",
    "    \"6dfs\": {\"antigen_chain\": \"D\"},\n",
    "    \"4p4k\": {\"antigen_chain\": \"B\"},\n",
    "    \"4may\": {\"antigen_chain\": \"D\"},\n",
    "}\n",
    "\n",
    "post_fasta_corrections = {\n",
    "    # 3tf7 has 1 tcrab pair bound together with linker\n",
    "    \"3tf7\": {\n",
    "        \"tcr_1_seq\": \"MGAQSVTQPDARVTVSEGASLQLRCKYSYSATPYLFWYVQYPRQGPQMLLKYYSGDPVVQGVNGFEAEFSKSDSSFHLRKASVHRSDSAVYFCAVSAKGTGSKLSFGKGAKLTVSP\",\n",
    "        \"tcr_2_seq\": \"SEAAVTQSPRNKVTVTGENVTLSCRQTNSHNYMYWYRQDTGHELRLIYYSYGAGNLQIGDVPDGYKATRTTQEDFFLTLESASPSQTSLYFCASSDAPGQLYFGEGSKLTVLELEHHHHHH\",\n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "for pdb_id, correction in pre_fasta_corrections.items():\n",
    "\n",
    "    for k, v in correction.items():\n",
    "        stcr_dat_phil = update_df_from_k_v(\n",
    "            stcr_dat_phil,\n",
    "            \"pdb\",\n",
    "            pdb_id,\n",
    "            k,\n",
    "            v,\n",
    "        )\n",
    "\n",
    "stcr_dat_phil = format_seqs(stcr_dat_phil)\n",
    "\n",
    "for pdb_id, correction in post_fasta_corrections.items():\n",
    "    for k, v in correction.items():\n",
    "        stcr_dat_phil = update_df_from_k_v(\n",
    "            stcr_dat_phil,\n",
    "            \"pdb\",\n",
    "            pdb_id,\n",
    "            k,\n",
    "            v,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Row STCRDab could not find\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclusion_pdb_date = get_pdb_date({\"pdb\": \"6l9l\"}).select(\"pdb_date\").item()\n",
    "\n",
    "exclusion = format_seqs(\n",
    "    phil_pdb.join(all_pdb, on=\"pdb\", how=\"anti\").with_columns(\n",
    "        Bchain=pl.lit(\"D\"),\n",
    "        Achain=pl.lit(\"C\"),\n",
    "        mhc_chain1=pl.lit(\"A\"),\n",
    "        mhc_chain2=pl.lit(None, dtype=pl.String),\n",
    "        antigen_chain=pl.lit(\"B\"),\n",
    "        mhc_1_species=pl.lit(\"mouse\"),\n",
    "        mhc_2_species=pl.lit(None, dtype=pl.String),\n",
    "        tcr_1_species=pl.lit(None, dtype=pl.String),\n",
    "        tcr_2_species=pl.lit(None, dtype=pl.String),\n",
    "        mhc_1_chain=pl.lit(\"heavy\"),\n",
    "        mhc_2_chain=pl.lit(None, dtype=pl.String),\n",
    "        cognate=pl.lit(True),\n",
    "        tcr_1_chain=pl.lit(\"alpha\"),\n",
    "        tcr_2_chain=pl.lit(\"beta\"),\n",
    "        pdb_date=exclusion_pdb_date,\n",
    "    )\n",
    ")\n",
    "\n",
    "phil_pdb = pl.concat(\n",
    "    [\n",
    "        exclusion,\n",
    "        stcr_dat_phil.select(pl.exclude(\"mhc_class_right\")),\n",
    "    ],\n",
    "    how=\"vertical_relaxed\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "phil_pdb_II = (\n",
    "    phil_pdb.filter(pl.col(\"mhc_class\") == \"II\")\n",
    "    .with_columns(pl.col(\"mhc\").str.split(\",\").alias(\"split_parts\"))\n",
    "    .with_columns(\n",
    "        pl.when(pl.col(\"split_parts\").list.len() == 2)\n",
    "        .then(\n",
    "            pl.struct(\n",
    "                pl.col(\"split_parts\")\n",
    "                .list.get(0, null_on_oob=True)\n",
    "                .alias(\"mhc_1_name\"),\n",
    "                pl.col(\"split_parts\")\n",
    "                .list.get(1, null_on_oob=True)\n",
    "                .alias(\"mhc_2_name\"),\n",
    "            )\n",
    "        )\n",
    "        .otherwise(\n",
    "            pl.struct(\n",
    "                pl.lit(None).alias(\"mhc_1_name\"),\n",
    "                pl.col(\"split_parts\").list.get(0).alias(\"mhc_2_name\"),\n",
    "            )\n",
    "        )\n",
    "        .alias(\"mhc_struct\")\n",
    "    )\n",
    "    .unnest(\"mhc_struct\")\n",
    ")\n",
    "\n",
    "phil_pdb_I = (\n",
    "    phil_pdb.filter(pl.col(\"mhc_class\") == \"I\")\n",
    "    .with_columns(pl.col(\"mhc\").str.split(\",\").alias(\"split_parts\"))\n",
    "    .with_columns(\n",
    "        pl.when(pl.col(\"split_parts\").list.len() == 2)\n",
    "        .then(\n",
    "            pl.struct(\n",
    "                pl.col(\"split_parts\")\n",
    "                .list.get(0, null_on_oob=True)\n",
    "                .alias(\"mhc_1_name\"),\n",
    "                pl.col(\"split_parts\")\n",
    "                .list.get(1, null_on_oob=True)\n",
    "                .alias(\"mhc_2_name\"),\n",
    "            )\n",
    "        )\n",
    "        .otherwise(\n",
    "            pl.struct(\n",
    "                pl.lit(\"B2M\").alias(\"mhc_2_name\"),\n",
    "                pl.col(\"split_parts\").list.get(0).alias(\"mhc_1_name\"),\n",
    "            )\n",
    "        )\n",
    "        .alias(\"mhc_struct\")\n",
    "    )\n",
    "    .unnest(\"mhc_struct\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "phil_pdb = pl.concat([phil_pdb_I, phil_pdb_II])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 0/130 [00:00<?, ?it/s]/tmp/ipykernel_3004030/938838702.py:236: UserWarning: Peptide found in TCR 2 sequence for PDB 4may at position 4\n",
      "  warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 4p4k at position 1\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 6dfs at position 3\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 6mnn at position 2\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:236: UserWarning: Peptide found in TCR 2 sequence for PDB 4grl at position 5\n",
      "  warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 3c60 at position 1\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 6dfx at position 1\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 3rdt at position 1\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 3c5z at position 1\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 6dfw at position 3\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 3o6f at position 1\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:236: UserWarning: Peptide found in TCR 2 sequence for PDB 3pl6 at position 5\n",
      "  warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 6bga at position 2\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:236: UserWarning: Peptide found in TCR 2 sequence for PDB 4p2q at position 3\n",
      "  warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:236: UserWarning: Peptide found in TCR 2 sequence for PDB 4p2r at position 3\n",
      "  warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 4p2o at position 5\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:224: UserWarning: Peptide found in MHC 2 sequence for PDB 5ks9 at position 2\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "/tmp/ipykernel_3004030/938838702.py:236: UserWarning: Peptide found in TCR 2 sequence for PDB 3qib at position 7\n",
      "  warnings.warn(f\"Peptide found in TCR 2 sequence for PDB {row['pdb']} at position {row['tcr_2_seq'].index(row['peptide'])}\")\n",
      "Processing rows: 100%|██████████| 130/130 [00:00<00:00, 3213.67it/s]\n"
     ]
    }
   ],
   "source": [
    "from tcr_format_parsers.common.TriadUtils import FORMAT_COLS, generate_job_name\n",
    "\n",
    "\n",
    "phil_pdb = generate_job_name(phil_pdb)\n",
    "\n",
    "phil_pdb = phil_pdb.select(\n",
    "    FORMAT_COLS\n",
    "    + [\n",
    "        \"pdb\",\n",
    "        \"pdb_date\",\n",
    "        \"va\",\n",
    "        \"ja\",\n",
    "        \"cdr3a\",\n",
    "        \"vb\",\n",
    "        \"jb\",\n",
    "        \"cdr3b\",\n",
    "        \"antigen_chain\",\n",
    "        \"mhc_chain1\",\n",
    "        \"mhc_chain2\",\n",
    "        \"Achain\",\n",
    "        \"Bchain\",\n",
    "        \"organism\",\n",
    "        \"cdr_rmsd\",\n",
    "        \"cdr_rmsd_af2_full\",\n",
    "        \"cdr_rmsd_af2_trim\",\n",
    "    ]\n",
    ")\n",
    "\n",
    "phil_pdb = serial_apply(phil_pdb, remove_peptide_from_chains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "phil_pdb.write_csv(\n",
    "    \"elife_replicate/phil_pdb.csv\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "phil_pdb = pl.read_csv(\n",
    "    \"elife_replicate/phil_pdb.csv\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    phil_pdb.with_columns(\n",
    "        [\n",
    "            pl.when(pl.col(\"mhc_class\") == \"I\")\n",
    "            .then(pl.lit(1))\n",
    "            .otherwise(pl.lit(2))\n",
    "            .alias(\"mhc_class\"),\n",
    "            pl.lit([1, 2, 3, 4, 5]).alias(\"replicate\"),\n",
    "        ]\n",
    "    )\n",
    "    .explode(\"replicate\")\n",
    "    .with_columns(\n",
    "        (pl.col(\"pdb\") + \"_\" + pl.col(\"replicate\").cast(pl.Utf8)).alias(\n",
    "            \"pdbid\"\n",
    "        )\n",
    "    )\n",
    "    .select(\"pdbid\", \"mhc_class\", \"organism\")\n",
    "    .write_csv(\"elife_replicate/phil_pdb_tcrdock_by_seed.csv\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert replication inference CIFs into TCRDock format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 130/130 [00:27<00:00,  4.78it/s]\n",
      "Processing rows: 100%|██████████| 130/130 [00:30<00:00,  4.32it/s]\n",
      "Processing rows: 100%|██████████| 130/130 [00:30<00:00,  4.21it/s]\n",
      "Processing rows: 100%|██████████| 130/130 [00:31<00:00,  4.11it/s]\n",
      "Processing rows: 100%|██████████| 130/130 [00:31<00:00,  4.10it/s]\n",
      "Processing rows: 100%|██████████| 130/130 [00:31<00:00,  4.15it/s]\n"
     ]
    }
   ],
   "source": [
    "import MDAnalysis as mda\n",
    "from mdaf3.FeatureExtraction import *\n",
    "from mdaf3.AF3OutputParser import *\n",
    "import Bio.Align\n",
    "\n",
    "\n",
    "def tcrdock_format_cif(row, inference_path, output_path, seed=None):\n",
    "    af3_output = AF3Output(inference_path / row[\"job_name\"])\n",
    "\n",
    "    pred_u = af3_output.get_mda_universe(seed=seed)\n",
    "\n",
    "    # maybe mhc 2 seq is not included since it's implied B2m, so check\n",
    "    if row[\"mhc_class\"] == \"II\":\n",
    "        # mhc1, mhc2, pep, tcr1, tcr2\n",
    "        pred_segids = [\"B\", \"C\", \"A\", \"D\", \"E\"]\n",
    "        rename_segids = [\"A\", \"B\", \"C\", \"D\", \"E\"]\n",
    "    # TCRdock will remove B2M anyways\n",
    "    else:\n",
    "        pred_segids = [\"B\", \"A\", \"D\", \"E\"]\n",
    "        # for consistency with format already in tcrdock repo\n",
    "        rename_segids = [\"A\", \"B\", \"C\", \"D\"]\n",
    "\n",
    "    chain_us = []\n",
    "\n",
    "    for pred_segsel, rename_segid in zip(pred_segids, rename_segids):\n",
    "\n",
    "        pred_sel = pred_u.select_atoms(f\"segid {pred_segsel}\").atoms\n",
    "\n",
    "        chain_u = mda.Merge(pred_sel)\n",
    "        chain_u.segments.segids = rename_segid\n",
    "        chain_u.atoms.chainIDs = [rename_segid] * len(chain_u.atoms)\n",
    "        chain_us.append(chain_u.atoms)\n",
    "\n",
    "    new_u = mda.Merge(*chain_us)\n",
    "\n",
    "    if seed is not None:\n",
    "        suffix = f\"_{seed}\"\n",
    "    else:\n",
    "        suffix = \"\"\n",
    "\n",
    "    with mda.Writer(output_path / (row[\"pdb\"] + suffix + \".pdb\")) as W:\n",
    "\n",
    "        # u_new = mda.Universe.empty(\n",
    "        #     n_atoms, n_segments=n_segments, n_residues=n_residues\n",
    "        # )\n",
    "\n",
    "        # ordered_chains = sum(chain_sels)\n",
    "\n",
    "        # # for attr in [\"name\", \"type\", \"resname\"]:\n",
    "        # #     u_new.add_TopologyAttr(\"name\", ordered_chains.residues.names)\n",
    "\n",
    "        # # choose first altloc if mutliple present\n",
    "        # W.write(ordered_chains)\n",
    "\n",
    "        W.write(new_u.atoms)\n",
    "\n",
    "    # noop\n",
    "    return pl.DataFrame(row)\n",
    "\n",
    "\n",
    "output_path = Path(\"elife_replicate/tcrdock_inference_pdb\")\n",
    "inference_path = Path(\"elife_replicate/inference\")\n",
    "\n",
    "serial_apply(\n",
    "    phil_pdb, tcrdock_format_cif, inference_path, output_path, seed=None\n",
    ")\n",
    "\n",
    "for seed in [1, 2, 3, 4, 5]:\n",
    "    serial_apply(\n",
    "        phil_pdb, tcrdock_format_cif, inference_path, output_path, seed=seed\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. New triads contributed after AF3 cutoff\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "post_cutoff_pdb = all_pdb.filter(pl.col(\"pdb_date\") > cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "accept = [\n",
    "    \"8gom\",\n",
    "    \"8vd0\",\n",
    "    \"8trr\",\n",
    "    \"8wte\",\n",
    "    \"8vcy\",\n",
    "    \"8es9\",\n",
    "    \"8gon\",\n",
    "    \"8i5d\",\n",
    "    \"8i5c\",\n",
    "    \"8vcx\",\n",
    "    \"7q99\",\n",
    "    \"8eo8\",\n",
    "    \"8dnt\",\n",
    "    \"8enh\",\n",
    "    \"8ye4\",\n",
    "    \"8wul\",\n",
    "    \"8f5a\",\n",
    "    \"8vd2\",\n",
    "    \"7q9b\",\n",
    "    \"8en8\",\n",
    "    \"8pjg\",\n",
    "    \"7q9a\",\n",
    "]\n",
    "organism = [\"human\"] * len(accept)\n",
    "\n",
    "tmp_metadat = pl.DataFrame({\"pdb\": accept, \"organism\": organism})\n",
    "post_cutoff_pdb = post_cutoff_pdb.join(tmp_metadat, on=\"pdb\", how=\"inner\")\n",
    "\n",
    "pre_fasta_corrections = {\n",
    "    \"8vd0\": {\"antigen_chain\": \"C\"},\n",
    "}\n",
    "\n",
    "post_fasta_corrections = {\n",
    "    \"8vd0\": {\"peptide\": \"GQVELGGGNAVEVCKG\"},\n",
    "    \"7q9b\": {\n",
    "        \"mhc_chain1\": \"FFF\",\n",
    "        \"mhc_chain2\": \"GGG\",\n",
    "        \"Achain\": \"III\",\n",
    "        \"Bchain\": \"JJJ\",\n",
    "        \"antigen_chain\": \"HHH\",\n",
    "    },\n",
    "}\n",
    "\n",
    "for pdb_id, correction in pre_fasta_corrections.items():\n",
    "\n",
    "    for k, v in correction.items():\n",
    "        post_cutoff_pdb = update_df_from_k_v(\n",
    "            post_cutoff_pdb,\n",
    "            \"pdb\",\n",
    "            pdb_id,\n",
    "            k,\n",
    "            v,\n",
    "        )\n",
    "\n",
    "\n",
    "post_cutoff_pdb = format_seqs(post_cutoff_pdb).rename(\n",
    "    {\"peptide_seq\": \"peptide\"}\n",
    ")\n",
    "\n",
    "for pdb_id, correction in post_fasta_corrections.items():\n",
    "    for k, v in correction.items():\n",
    "        post_cutoff_pdb = update_df_from_k_v(\n",
    "            post_cutoff_pdb,\n",
    "            \"pdb\",\n",
    "            pdb_id,\n",
    "            k,\n",
    "            v,\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Infer the HLA alleles- not used downstream, so does not have to be perfect\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 22/22 [05:46<00:00, 15.74s/it]\n"
     ]
    }
   ],
   "source": [
    "from tcr_format_parsers.common.MHCCodeConverter import HLASequenceDBConverter\n",
    "\n",
    "conv = HLASequenceDBConverter(IMGT_HLA_PATH)\n",
    "\n",
    "post_cutoff_pdb = serial_apply(post_cutoff_pdb, infer_correct_mhc, conv, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tcr_format_parsers.common.TriadUtils import generate_job_name, FORMAT_COLS\n",
    "\n",
    "post_cutoff_pdb = generate_job_name(post_cutoff_pdb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 0/22 [00:00<?, ?it/s]/tmp/ipykernel_4190352/3855872097.py:227: UserWarning: Peptide found in MHC 2 sequence for PDB 8vd0 at position 0\n",
      "  warnings.warn(f\"Peptide found in MHC 2 sequence for PDB {row['pdb']} at position {row['mhc_2_seq'].index(row['peptide'])}\")\n",
      "Processing rows: 100%|██████████| 22/22 [00:00<00:00, 1924.72it/s]\n"
     ]
    }
   ],
   "source": [
    "post_cutoff_pdb = post_cutoff_pdb.select(\n",
    "    FORMAT_COLS\n",
    "    + [\n",
    "        \"pdb\",\n",
    "        \"pdb_date\",\n",
    "        \"antigen_chain\",\n",
    "        \"mhc_chain1\",\n",
    "        \"mhc_chain2\",\n",
    "        \"Achain\",\n",
    "        \"Bchain\",\n",
    "        \"organism\",\n",
    "    ]\n",
    ")\n",
    "\n",
    "post_cutoff_pdb = serial_apply(post_cutoff_pdb, remove_peptide_from_chains)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 22/22 [00:22<00:00,  1.02s/it]\n"
     ]
    }
   ],
   "source": [
    "post_cutoff_pdb = serial_apply(\n",
    "    post_cutoff_pdb, download_pdb, Path(\"post_training/struct\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "post_cutoff_pdb.write_csv(\n",
    "    \"post_training/post_cutoff_pdb.csv\",\n",
    ")\n",
    "\n",
    "# also write into TCRDock CSV format\n",
    "# this was provided for us in the original paper, but here we have to write it\n",
    "\n",
    "post_cutoff_pdb.with_columns(\n",
    "    pl.when(pl.col(\"mhc_class\") == \"I\")\n",
    "    .then(pl.lit(1))\n",
    "    .otherwise(pl.lit(2))\n",
    "    .alias(\"mhc_class\"),\n",
    "    pl.col(\"pdb\").alias(\"pdbid\"),\n",
    ").select(\"pdbid\", \"mhc_class\", \"organism\").write_csv(\n",
    "    \"post_training/post_cutoff_pdb_tcrdock.csv\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "\n",
    "post_cutoff_pdb = pl.read_csv(\n",
    "    \"post_training/post_cutoff_pdb.csv\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    post_cutoff_pdb.with_columns(\n",
    "        [\n",
    "            pl.when(pl.col(\"mhc_class\") == \"I\")\n",
    "            .then(pl.lit(1))\n",
    "            .otherwise(pl.lit(2))\n",
    "            .alias(\"mhc_class\"),\n",
    "            pl.lit([1, 2, 3, 4, 5]).alias(\"replicate\"),\n",
    "        ]\n",
    "    )\n",
    "    .explode(\"replicate\")\n",
    "    .with_columns(\n",
    "        (pl.col(\"pdb\") + \"_\" + pl.col(\"replicate\").cast(pl.Utf8)).alias(\n",
    "            \"pdbid\"\n",
    "        )\n",
    "    )\n",
    "    .select(\"pdbid\", \"mhc_class\", \"organism\")\n",
    "    .write_csv(\"post_training/post_cutoff_pdb_tcrdock_by_seed.csv\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert post-AF3 training date inference CIFs into TCRDock format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 22/22 [00:04<00:00,  4.44it/s]\n",
      "Processing rows: 100%|██████████| 22/22 [00:05<00:00,  4.02it/s]\n",
      "Processing rows: 100%|██████████| 22/22 [00:05<00:00,  4.14it/s]\n",
      "Processing rows: 100%|██████████| 22/22 [00:05<00:00,  4.08it/s]\n",
      "Processing rows: 100%|██████████| 22/22 [00:05<00:00,  3.93it/s]\n",
      "Processing rows: 100%|██████████| 22/22 [00:06<00:00,  3.54it/s]\n"
     ]
    }
   ],
   "source": [
    "import MDAnalysis as mda\n",
    "from mdaf3.FeatureExtraction import *\n",
    "from mdaf3.AF3OutputParser import *\n",
    "import Bio.Align\n",
    "\n",
    "\n",
    "# def tcrdock_format_cif(row, inference_path, output_path, seed=None):\n",
    "#     af3_output = AF3Output(inference_path / row[\"job_name\"], seed=seed)\n",
    "\n",
    "#     pred_u = af3_output.get_mda_universe()\n",
    "\n",
    "#     # maybe mhc 2 seq is not included since it's implied B2m, so check\n",
    "#     if row[\"mhc_class\"] == \"II\":\n",
    "#         # mhc1, mhc2, pep, tcr1, tcr2\n",
    "#         pred_segids = [\"B\", \"C\", \"A\", \"D\", \"E\"]\n",
    "#         rename_segids = [\"A\", \"B\", \"C\", \"D\", \"E\"]\n",
    "#     # TCRdock will remove B2M anyways\n",
    "#     else:\n",
    "#         pred_segids = [\"B\", \"A\", \"D\", \"E\"]\n",
    "#         # for consistency with format already in tcrdock repo\n",
    "#         rename_segids = [\"A\", \"B\", \"C\", \"D\"]\n",
    "\n",
    "#     chain_us = []\n",
    "\n",
    "#     for pred_segsel, rename_segid in zip(pred_segids, rename_segids):\n",
    "\n",
    "#         pred_sel = pred_u.select_atoms(f\"segid {pred_segsel}\").atoms\n",
    "\n",
    "#         chain_u = mda.Merge(pred_sel)\n",
    "#         chain_u.segments.segids = rename_segid\n",
    "#         chain_u.atoms.chainIDs = [rename_segid] * len(chain_u.atoms)\n",
    "#         chain_us.append(chain_u.atoms)\n",
    "\n",
    "#     new_u = mda.Merge(*chain_us)\n",
    "\n",
    "#     with mda.Writer(output_path / (row[\"pdb\"] + \".pdb\")) as W:\n",
    "\n",
    "#         # u_new = mda.Universe.empty(\n",
    "#         #     n_atoms, n_segments=n_segments, n_residues=n_residues\n",
    "#         # )\n",
    "\n",
    "#         # ordered_chains = sum(chain_sels)\n",
    "\n",
    "#         # # for attr in [\"name\", \"type\", \"resname\"]:\n",
    "#         # #     u_new.add_TopologyAttr(\"name\", ordered_chains.residues.names)\n",
    "\n",
    "#         # # choose first altloc if mutliple present\n",
    "#         # W.write(ordered_chains)\n",
    "\n",
    "#         W.write(new_u.atoms)\n",
    "\n",
    "#     # noop\n",
    "#     return pl.DataFrame(row)\n",
    "\n",
    "\n",
    "output_path = Path(\"post_training/tcrdock_inference_pdb\")\n",
    "inference_path = Path(\"post_training/inference\")\n",
    "\n",
    "\n",
    "serial_apply(\n",
    "    post_cutoff_pdb,\n",
    "    tcrdock_format_cif,\n",
    "    inference_path,\n",
    "    output_path,\n",
    "    seed=None,\n",
    ")\n",
    "\n",
    "for seed in [1, 2, 3, 4, 5]:\n",
    "    serial_apply(\n",
    "        post_cutoff_pdb,\n",
    "        tcrdock_format_cif,\n",
    "        inference_path,\n",
    "        output_path,\n",
    "        seed=seed,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert post-AF3 training date raw PDBs into TCRDock format\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|▍         | 1/22 [00:00<00:08,  2.60it/s]/home/lwoods/workspace/mdanalysis/package/MDAnalysis/coordinates/PDB.py:1154: UserWarning: Found no information for attr: 'formalcharges' Using default value of '0'\n",
      "  warnings.warn(\"Found no information for attr: '{}'\"\n",
      "Processing rows:  36%|███▋      | 8/22 [00:03<00:06,  2.31it/s]/home/lwoods/workspace/mdanalysis/package/MDAnalysis/coordinates/PDB.py:453: UserWarning: 1 A^3 CRYST1 record, this is usually a placeholder. Unit cell dimensions will be set to None.\n",
      "  warnings.warn(\"1 A^3 CRYST1 record,\"\n",
      "Processing rows: 100%|██████████| 22/22 [00:09<00:00,  2.30it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (22, 26)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>job_name</th><th>cognate</th><th>peptide</th><th>mhc_class</th><th>mhc_1_chain</th><th>mhc_1_species</th><th>mhc_1_name</th><th>mhc_1_seq</th><th>mhc_2_chain</th><th>mhc_2_species</th><th>mhc_2_name</th><th>mhc_2_seq</th><th>tcr_1_chain</th><th>tcr_1_species</th><th>tcr_1_seq</th><th>tcr_2_chain</th><th>tcr_2_species</th><th>tcr_2_seq</th><th>pdb</th><th>pdb_date</th><th>antigen_chain</th><th>mhc_chain1</th><th>mhc_chain2</th><th>Achain</th><th>Bchain</th><th>organism</th></tr><tr><td>str</td><td>bool</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;b4a6153b3bd2df0309a16f643d9be8…</td><td>true</td><td>&quot;EAAGIGILTV&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>null</td><td>&quot;A*02:01&quot;</td><td>&quot;GSHSMRYFFTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>null</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>null</td><td>&quot;QKEVEQNSGPLSVPEGAIASLNCTYSDRGS…</td><td>&quot;beta&quot;</td><td>null</td><td>&quot;NAGVTQTPKFQVLKTGQSMTLQCAQDMNHE…</td><td>&quot;7q9b&quot;</td><td>&quot;2023-02-22T00:00:00.000000+000…</td><td>&quot;HHH&quot;</td><td>&quot;FFF&quot;</td><td>&quot;GGG&quot;</td><td>&quot;III&quot;</td><td>&quot;JJJ&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;39bf4e6e65f6b042d9340128ecd790…</td><td>true</td><td>&quot;GQVELGGGNAVEVCKG&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*03:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYSHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*03:02&quot;</td><td>&quot;GSGGSRDSPEDFVYQFKGMCYFTNGTERVR…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVTQTPRYLIKTRGQQVTLSCSPISGHRSV…</td><td>&quot;8vd0&quot;</td><td>&quot;2024-08-07T00:00:00.000000+000…</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;62f1b197f4b0555c27ac38c5496ffd…</td><td>true</td><td>&quot;NLSALGIFST&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*02:01&quot;</td><td>&quot;GSHSMRYFFTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;EVEQNSGPLSVPEGAIASLNCTYSDRGSQS…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;SQTIHQWPATLVQPVGSPLSLECTVEGTSN…</td><td>&quot;7q99&quot;</td><td>&quot;2023-02-22T00:00:00.000000+000…</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;b093517b1cc89f77734e54d40b30ca…</td><td>true</td><td>&quot;LPFEKSTIM&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;B*35:01&quot;</td><td>&quot;GSHSMRYFYTAMSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;SAGENVEQHPSTLSVQEGDSAVIKCTYSDS…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;AVVSQHPSRVICKSGTSVKIECRSLDFQAT…</td><td>&quot;8enh&quot;</td><td>&quot;2024-03-27T00:00:00.000000+000…</td><td>&quot;H&quot;</td><td>&quot;F&quot;</td><td>&quot;G&quot;</td><td>&quot;I&quot;</td><td>&quot;J&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;3dad0fd6a6e0ce25e4106fda13a57a…</td><td>true</td><td>&quot;VVGAVGVGK&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*11:100&quot;</td><td>&quot;GSHSMRYFYTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;QQKVQQSPESLIVPEGGMASLNCTSSDRNV…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;EAAVTQSPRNKVAVTGGKVTLSCNQTNNHN…</td><td>&quot;8i5c&quot;</td><td>&quot;2023-08-23T00:00:00.000000+000…</td><td>&quot;M&quot;</td><td>&quot;K&quot;</td><td>&quot;L&quot;</td><td>&quot;N&quot;</td><td>&quot;O&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;58484be6ef5feee9f02e558e069f54…</td><td>true</td><td>&quot;LPFDKATIM&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;B*35:01&quot;</td><td>&quot;GSHSMRYFYTAMSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;SAGENVEQHPSTLSVQEGDSAVIKCTYSDS…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;AVVSQHPSRVICKSGTSVKIECRSLDFQAT…</td><td>&quot;8eo8&quot;</td><td>&quot;2024-03-27T00:00:00.000000+000…</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;59acb5477505053567be60b57e5f3b…</td><td>true</td><td>&quot;LLLDRLNQL&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*02:01&quot;</td><td>&quot;MGSHSMRYFFTSVSRPGRGEPRFIAVGYVD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;KEVEQNSGPLSVPEGAIASLNCTYSDRGSQ…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GAGVSQSPSNKVTEKGKDVELRCDPISGHT…</td><td>&quot;8dnt&quot;</td><td>&quot;2023-07-19T00:00:00.000000+000…</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;F&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;13f7a9410810562e44920fa0413159…</td><td>true</td><td>&quot;VVGAVGVGK&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*11:01&quot;</td><td>&quot;GSHSMRYFYTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;QQKVQQSPESLIVPEGGMASLNCTSSDRNV…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;KIIQKPKYLVAVTGSEKILICEQYLGHNAM…</td><td>&quot;8wte&quot;</td><td>&quot;2024-05-01T00:00:00.000000+000…</td><td>&quot;J&quot;</td><td>&quot;H&quot;</td><td>&quot;I&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;e6da04b0f02efa07a5e1795130328a…</td><td>true</td><td>&quot;GQVELGGGSSPETCI&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*03:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYSHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*03:02&quot;</td><td>&quot;RDSPEDFVYQFKGMCYFTNGTERVRLVTRY…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVTQTPRYLIKTRGQQVTLSCSPISGHRSV…</td><td>&quot;8vcy&quot;</td><td>&quot;2024-08-07T00:00:00.000000+000…</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;612142b21e38127bab9e363c4d0eb8…</td><td>true</td><td>&quot;RLQSLQIYV&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*02:01&quot;</td><td>&quot;MGSHSMRYFFTSVSRPGRGEPRFIAVGYVD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MAQTVTQSQPEMSVQEAETVTLSCTYDTSE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;MDAGVIQSPRHEVTEMGQEVTLRCKPISGH…</td><td>&quot;8gon&quot;</td><td>&quot;2023-03-01T00:00:00.000000+000…</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (22, 26)\n",
       "┌───────────────┬─────────┬──────────────┬───────────┬───┬────────────┬────────┬────────┬──────────┐\n",
       "│ job_name      ┆ cognate ┆ peptide      ┆ mhc_class ┆ … ┆ mhc_chain2 ┆ Achain ┆ Bchain ┆ organism │\n",
       "│ ---           ┆ ---     ┆ ---          ┆ ---       ┆   ┆ ---        ┆ ---    ┆ ---    ┆ ---      │\n",
       "│ str           ┆ bool    ┆ str          ┆ str       ┆   ┆ str        ┆ str    ┆ str    ┆ str      │\n",
       "╞═══════════════╪═════════╪══════════════╪═══════════╪═══╪════════════╪════════╪════════╪══════════╡\n",
       "│ b4a6153b3bd2d ┆ true    ┆ EAAGIGILTV   ┆ I         ┆ … ┆ GGG        ┆ III    ┆ JJJ    ┆ human    │\n",
       "│ f0309a16f643d ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 9be8…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 39bf4e6e65f6b ┆ true    ┆ GQVELGGGNAVE ┆ II        ┆ … ┆ C          ┆ D      ┆ E      ┆ human    │\n",
       "│ 042d9340128ec ┆         ┆ VCKG         ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ d790…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 62f1b197f4b05 ┆ true    ┆ NLSALGIFST   ┆ I         ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ 55c27ac38c549 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 6ffd…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ b093517b1cc89 ┆ true    ┆ LPFEKSTIM    ┆ I         ┆ … ┆ G          ┆ I      ┆ J      ┆ human    │\n",
       "│ f77734e54d40b ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 30ca…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 3dad0fd6a6e0c ┆ true    ┆ VVGAVGVGK    ┆ I         ┆ … ┆ L          ┆ N      ┆ O      ┆ human    │\n",
       "│ e25e4106fda13 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ a57a…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ …             ┆ …       ┆ …            ┆ …         ┆ … ┆ …          ┆ …      ┆ …      ┆ …        │\n",
       "│ 58484be6ef5fe ┆ true    ┆ LPFDKATIM    ┆ I         ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ ee9f02e558e06 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 9f54…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 59acb54775050 ┆ true    ┆ LLLDRLNQL    ┆ I         ┆ … ┆ F          ┆ A      ┆ B      ┆ human    │\n",
       "│ 53567be60b57e ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 5f3b…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 13f7a94108105 ┆ true    ┆ VVGAVGVGK    ┆ I         ┆ … ┆ I          ┆ A      ┆ B      ┆ human    │\n",
       "│ 62e44920fa041 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 3159…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ e6da04b0f02ef ┆ true    ┆ GQVELGGGSSPE ┆ II        ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ a07a5e1795130 ┆         ┆ TCI          ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 328a…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 612142b21e381 ┆ true    ┆ RLQSLQIYV    ┆ I         ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ 27bab9e363c4d ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 0eb8…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "└───────────────┴─────────┴──────────────┴───────────┴───┴────────────┴────────┴────────┴──────────┘"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import MDAnalysis as mda\n",
    "from mdaf3.FeatureExtraction import *\n",
    "from mdaf3.AF3OutputParser import *\n",
    "import Bio.Align\n",
    "\n",
    "\n",
    "def align_residue_groups(true_residues, pred_residues, strict=False):\n",
    "    aligner = Bio.Align.PairwiseAligner(mode=\"global\")\n",
    "\n",
    "    # Alignment[0] is best alignment\n",
    "    seq_aln = aligner.align(\n",
    "        true_residues.sequence(format=\"string\"),\n",
    "        pred_residues.sequence(format=\"string\"),\n",
    "    )[0]\n",
    "\n",
    "    true_aln = seq_aln[0]\n",
    "    pred_aln = seq_aln[1]\n",
    "\n",
    "    true_resindices = true_residues.resindices\n",
    "    true_res_boolmask = np.full(true_resindices.shape, False, dtype=np.bool)\n",
    "    pred_resindices = pred_residues.resindices\n",
    "    pred_res_boolmask = np.full(pred_resindices.shape, False, dtype=np.bool)\n",
    "\n",
    "    true_ptr = 0\n",
    "    pred_ptr = 0\n",
    "    for i in range(len(true_aln)):\n",
    "        if true_aln[i] == \"-\":\n",
    "            pred_ptr += 1\n",
    "            continue\n",
    "        elif pred_aln[i] == \"-\":\n",
    "            true_ptr += 1\n",
    "            continue\n",
    "        elif strict and pred_aln[i] != true_aln[i]:\n",
    "            raise ValueError(\"Sequences cannot align without mutations\")\n",
    "        else:\n",
    "            true_res_boolmask[true_ptr] = True\n",
    "            pred_res_boolmask[pred_ptr] = True\n",
    "            true_ptr += 1\n",
    "            pred_ptr += 1\n",
    "\n",
    "    return true_residues[true_res_boolmask], pred_residues[pred_res_boolmask]\n",
    "\n",
    "\n",
    "def get_true_mda_universe(pdb_id, root_path):\n",
    "    # Favor PDB since it doesn't have multiple residue with same ID issue\n",
    "    if (root_path / (pdb_id + \".pdb\")).exists():\n",
    "        suffix = \".pdb\"\n",
    "\n",
    "    else:\n",
    "        suffix = \".cif\"\n",
    "\n",
    "    return mda.Universe((root_path / (pdb_id + suffix)).as_posix())\n",
    "\n",
    "\n",
    "def tcrdock_format_pdb(row, struct_path, inference_path, output_path):\n",
    "    af3_output = AF3Output(inference_path / row[\"job_name\"])\n",
    "\n",
    "    pred_u = af3_output.get_mda_universe()\n",
    "\n",
    "    true_u = get_true_mda_universe(row[\"pdb\"], struct_path)\n",
    "\n",
    "    if row[\"mhc_class\"] == \"II\":\n",
    "        # mhc1, mhc2, pep, tcr1, tcr2\n",
    "        pred_segids = [\"B\", \"C\", \"A\", \"D\", \"E\"]\n",
    "        true_segids = [\n",
    "            row[\"mhc_chain1\"],\n",
    "            row[\"mhc_chain2\"],\n",
    "            row[\"antigen_chain\"],\n",
    "            row[\"Achain\"],\n",
    "            row[\"Bchain\"],\n",
    "        ]\n",
    "        rename_segids = [\"A\", \"B\", \"C\", \"D\", \"E\"]\n",
    "    # TCRdock will remove B2M anyways\n",
    "    else:\n",
    "        pred_segids = [\"B\", \"A\", \"D\", \"E\"]\n",
    "        true_segids = [\n",
    "            row[\"mhc_chain1\"],\n",
    "            row[\"antigen_chain\"],\n",
    "            row[\"Achain\"],\n",
    "            row[\"Bchain\"],\n",
    "        ]\n",
    "        rename_segids = [\"A\", \"B\", \"C\", \"D\"]\n",
    "\n",
    "    # seg_map = {true_segids[i]: pred_segids[i] for i in range(len(true_segid))}\n",
    "\n",
    "    chain_us = []\n",
    "\n",
    "    for true_segsel, pred_segsel, rename_segid in zip(\n",
    "        true_segids, pred_segids, rename_segids\n",
    "    ):\n",
    "        true_sel = true_u.select_atoms(\n",
    "            f\"segid {true_segsel} and name CA and record_type ATOM\"\n",
    "        ).residues\n",
    "\n",
    "        pred_sel = pred_u.select_atoms(\n",
    "            f\"segid {pred_segsel} and name CA and record_type ATOM\"\n",
    "        ).residues\n",
    "\n",
    "        true_res, pred_res = align_residue_groups(\n",
    "            true_sel,\n",
    "            pred_sel,\n",
    "        )\n",
    "\n",
    "        chain_u = mda.Merge(\n",
    "            true_res.atoms.select_atoms(\n",
    "                \"record_type ATOM and (altloc A or not altloc [!?])\"\n",
    "            )\n",
    "        )\n",
    "        chain_u.segments.segids = rename_segid\n",
    "        chain_u.atoms.chainIDs = [rename_segid] * len(chain_u.atoms)\n",
    "\n",
    "        chain_us.append(chain_u.atoms)\n",
    "\n",
    "    new_u = mda.Merge(*chain_us)\n",
    "\n",
    "    with mda.Writer(output_path / (row[\"pdb\"] + \".pdb\")) as W:\n",
    "\n",
    "        W.write(new_u.atoms)\n",
    "\n",
    "    # noop\n",
    "    return pl.DataFrame(row)\n",
    "\n",
    "\n",
    "output_path = Path(\"post_training/tcrdock_struct_pdb\")\n",
    "struct_path = Path(\"post_training/struct\")\n",
    "inference_path = Path(\"post_training/inference\")\n",
    "\n",
    "serial_apply(\n",
    "    post_cutoff_pdb,\n",
    "    tcrdock_format_pdb,\n",
    "    struct_path,\n",
    "    inference_path,\n",
    "    output_path,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "phil_pdb = pl.read_csv(\n",
    "    \"elife_replicate/phil_pdb.csv\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 130/130 [00:06<00:00, 19.01it/s]\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime, timezone\n",
    "\n",
    "cutoff = pl.lit(datetime(2023, 1, 12, tzinfo=timezone.utc))\n",
    "\n",
    "phil_pdb = serial_apply(\n",
    "    phil_pdb,\n",
    "    get_pdb_date,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (21, 35)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>job_name</th><th>cognate</th><th>peptide</th><th>mhc_class</th><th>mhc_1_chain</th><th>mhc_1_species</th><th>mhc_1_name</th><th>mhc_1_seq</th><th>mhc_2_chain</th><th>mhc_2_species</th><th>mhc_2_name</th><th>mhc_2_seq</th><th>tcr_1_chain</th><th>tcr_1_species</th><th>tcr_1_seq</th><th>tcr_2_chain</th><th>tcr_2_species</th><th>tcr_2_seq</th><th>pdb</th><th>pdb_date</th><th>va</th><th>ja</th><th>cdr3a</th><th>vb</th><th>jb</th><th>cdr3b</th><th>antigen_chain</th><th>mhc_chain1</th><th>mhc_chain2</th><th>Achain</th><th>Bchain</th><th>organism</th><th>cdr_rmsd</th><th>cdr_rmsd_af2_full</th><th>cdr_rmsd_af2_trim</th></tr><tr><td>str</td><td>bool</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>datetime[μs, UTC]</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td></tr></thead><tbody><tr><td>&quot;00d671d629eff4094fee1be3c7f4da…</td><td>true</td><td>&quot;SPSYAYHQF&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;mouse&quot;</td><td>&quot;H2Ld&quot;</td><td>&quot;GPHSMRYYETATSRRGLGEPRYTSVGYVDD…</td><td>null</td><td>null</td><td>&quot;B2M&quot;</td><td>null</td><td>&quot;alpha&quot;</td><td>null</td><td>&quot;AKTTQPDSMESTEGETVHLPCSHATISGNE…</td><td>&quot;beta&quot;</td><td>null</td><td>&quot;AVTQSPRNKVTVTGGNVTLSCRQTNSHNYM…</td><td>&quot;6l9l&quot;</td><td>2019-11-10 00:00:00 UTC</td><td>&quot;TRAV21/DV12*01&quot;</td><td>&quot;TRAJ58*01&quot;</td><td>&quot;CILQGTGSKLSF&quot;</td><td>&quot;TRBV13-1*02&quot;</td><td>&quot;TRBJ2-7*01&quot;</td><td>&quot;CASSDGDYEQYF&quot;</td><td>&quot;B&quot;</td><td>&quot;A&quot;</td><td>null</td><td>&quot;C&quot;</td><td>&quot;D&quot;</td><td>&quot;mouse&quot;</td><td>4.2</td><td>11.69</td><td>26.58</td></tr><tr><td>&quot;66588fb6f41c42796a59e21f801dcc…</td><td>true</td><td>&quot;RLQSLQTYV&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*02:01&quot;</td><td>&quot;GSHSMRYFFTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;QRVTQPEKLLSVFKGAPVELKCNYSYSGSP…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVAQSPRYKIIEKRQSVAFWCNPISGHATL…</td><td>&quot;7n1e&quot;</td><td>2021-05-27 00:00:00 UTC</td><td>&quot;TRAV16*01&quot;</td><td>&quot;TRAJ39*01&quot;</td><td>&quot;CALSGFNNAGNMLTF&quot;</td><td>&quot;TRBV11-2*01&quot;</td><td>&quot;TRBJ2-3*01&quot;</td><td>&quot;CASSLGGAGGADTQYF&quot;</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td><td>8.07</td><td>22.91</td><td>26.8</td></tr><tr><td>&quot;c874eda106179b27d61348b7557431…</td><td>true</td><td>&quot;SLLMWITQV&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*02:01&quot;</td><td>&quot;MGSHSMRYFFTSVSRPGRGEPRFIAVGYVD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MIQRTPKIQVYSRHPAENGKSNFLNCYVSG…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MQKEVEQNSGPLSVPEGAIASLNCTYSDRG…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;MGAGVSQSPRYKVTKRGQDVALRCDPISGH…</td><td>&quot;6rp9&quot;</td><td>2019-05-14 00:00:00 UTC</td><td>&quot;TRAV12-2*01&quot;</td><td>&quot;TRAJ49*01&quot;</td><td>&quot;CALTRGPGNQFYF&quot;</td><td>&quot;TRBV7-6*01&quot;</td><td>&quot;TRBJ1-1*01&quot;</td><td>&quot;CASSSPGGVSTEAFF&quot;</td><td>&quot;H&quot;</td><td>&quot;F&quot;</td><td>&quot;G&quot;</td><td>&quot;I&quot;</td><td>&quot;J&quot;</td><td>&quot;human&quot;</td><td>5.67</td><td>23.64</td><td>20.27</td></tr><tr><td>&quot;1e2f70e91c8b2d204ab7d420da955a…</td><td>true</td><td>&quot;FEDLRVLSF&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;B*37:01&quot;</td><td>&quot;GSHSMRYFHTSVSRPGRGEPRFISVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;QPVQSPQAVILREGEDAVINCSSSKALYSV…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GITQSPKYLFRKEGQNVTLSCEQNLNHDAM…</td><td>&quot;6mtm&quot;</td><td>2018-10-19 00:00:00 UTC</td><td>&quot;TRAV30*01&quot;</td><td>&quot;TRAJ13*01&quot;</td><td>&quot;CGTERSGGYQKVTF&quot;</td><td>&quot;TRBV19*01&quot;</td><td>&quot;TRBJ1-1*01&quot;</td><td>&quot;CASSMSAMGTEAFF&quot;</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td><td>2.49</td><td>5.28</td><td>3.14</td></tr><tr><td>&quot;7f4b0f1dd8f99a68601d93e79d9738…</td><td>true</td><td>&quot;ASNENMETM&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;mouse&quot;</td><td>&quot;H2Db&quot;</td><td>&quot;MGAMAPRTLLLLLAAALAPTQTRAGPHSMR…</td><td>&quot;light&quot;</td><td>&quot;mouse&quot;</td><td>&quot;B2M&quot;</td><td>&quot;MARSVTLVFLVLVSLTGLYAIQKTPQIQVY…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;GDQVEQSPSALSLHEGTDSALRCNFTTTMR…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;DTTVKQNPRYKLARVGKPVNLICSQTMNHD…</td><td>&quot;7jwj&quot;</td><td>2020-08-25 00:00:00 UTC</td><td>&quot;TRAV4-4/DV10*01&quot;</td><td>&quot;TRAJ37*01&quot;</td><td>&quot;CAAVTGNTGKLIF&quot;</td><td>&quot;TRBV17*01&quot;</td><td>&quot;TRBJ1-1*01&quot;</td><td>&quot;CASSRGTIHSNTEVFF&quot;</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;mouse&quot;</td><td>8.6</td><td>25.94</td><td>16.93</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;cce720e42d01d1e8820318af338c7c…</td><td>true</td><td>&quot;SYYGPKTSPVQ&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;H2ABa&quot;</td><td>&quot;IEADHVGTYGISVYQSPGDIGQYTFEFDGD…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;H2ABb&quot;</td><td>&quot;GGGGSLVPRGSGGGGSERHFVYQFMGECYF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;MQQVRQSPQSLTVWEGETAILNCSYENSAF…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;AVTQSPRNKVAVTGGKVTLSCNQTNNHNNM…</td><td>&quot;6mnn&quot;</td><td>2018-10-02 00:00:00 UTC</td><td>&quot;TRAV14D-3/DV8*02&quot;</td><td>&quot;TRAJ52*01&quot;</td><td>&quot;CAASVTGANTGKLTF&quot;</td><td>&quot;TRBV13-2*01&quot;</td><td>&quot;TRBJ2-4*01&quot;</td><td>&quot;CASGDFWGDTLYF&quot;</td><td>&quot;D&quot;</td><td>&quot;C&quot;</td><td>&quot;D&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;mouse&quot;</td><td>3.83</td><td>3.89</td><td>4.58</td></tr><tr><td>&quot;da51f41cc36cf890da26880c6f14fe…</td><td>true</td><td>&quot;APFSEQEQPVL&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*02:01&quot;</td><td>&quot;MILNKALMLGALALTTVMSPCGGEDIVADH…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*02:01&quot;</td><td>&quot;MSWKKALRIPGGLRAATVTLMLSMLSTPVA…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKQEVTQIPAALSVPEGENLVLNCSFTDSA…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;MGVSQTPSNKVTEKGKYVELRCDPISGHTA…</td><td>&quot;6px6&quot;</td><td>2019-07-24 00:00:00 UTC</td><td>&quot;TRAV21*01&quot;</td><td>&quot;TRAJ31*01&quot;</td><td>&quot;CAVHTGARLMF&quot;</td><td>&quot;TRBV7-3*01&quot;</td><td>&quot;TRBJ2-3*01&quot;</td><td>&quot;CASSHGASTDTQYF&quot;</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td><td>3.9</td><td>2.81</td><td>3.33</td></tr><tr><td>&quot;b05a70ac15dcec9c3125e3649535e8…</td><td>true</td><td>&quot;APMPMPELPYP&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*05:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYTHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*02:01&quot;</td><td>&quot;GGSGASRDSPEDFVYQFKGMCYFTNGTERV…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;QSVTQPDIHITVSEGASLELRCNYSYGATP…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;HMGVTQSPTHLIKTRGQQVTLRCSPISGHK…</td><td>&quot;6u3n&quot;</td><td>2019-08-22 00:00:00 UTC</td><td>&quot;TRAV8-3*01&quot;</td><td>&quot;TRAJ33*01&quot;</td><td>&quot;CAVGAGSNYQLIW&quot;</td><td>&quot;TRBV5-5*01&quot;</td><td>&quot;TRBJ2-1*01&quot;</td><td>&quot;CASSLEGQGASEQFF&quot;</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td><td>3.2</td><td>4.52</td><td>4.53</td></tr><tr><td>&quot;08a74d65f9e5f41a5ec67c0accfde9…</td><td>true</td><td>&quot;VVQSELPYPEG&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*05:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYTHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*02:01&quot;</td><td>&quot;GGSGASRDSPEDFVYQFKGMCYFTNGTERV…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;MGVSQSPSNKVTEKGKDVELRCDPISGHTA…</td><td>&quot;6u3o&quot;</td><td>2019-08-22 00:00:00 UTC</td><td>&quot;TRAV26-1*01&quot;</td><td>&quot;TRAJ54*01&quot;</td><td>&quot;CIAFQGAQKLVF&quot;</td><td>&quot;TRBV7-2*01&quot;</td><td>&quot;TRBJ2-3*01&quot;</td><td>&quot;CASSFRALAADTQYF&quot;</td><td>&quot;I&quot;</td><td>&quot;E&quot;</td><td>&quot;F&quot;</td><td>&quot;G&quot;</td><td>&quot;H&quot;</td><td>&quot;human&quot;</td><td>3.03</td><td>20.29</td><td>5.43</td></tr><tr><td>&quot;a002927670058c8f7dc0ebffd01939…</td><td>true</td><td>&quot;APFSEQEQPVL&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*02:01&quot;</td><td>&quot;MILNKALMLGALALTTVMSPCGGEDIVADH…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*02:01&quot;</td><td>&quot;MSWKKALRIPGGLRAATVTLMLSMLSTPVA…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MNSVTQMEGPVTLSEEAFLTINCTYTATGY…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;MGVAQSPRYKIIEKRQSVAFWCNPISGHAT…</td><td>&quot;6py2&quot;</td><td>2019-07-28 00:00:00 UTC</td><td>&quot;TRAV9-2*04&quot;</td><td>&quot;TRAJ57*01&quot;</td><td>&quot;CASPQGGSEKLVF&quot;</td><td>&quot;TRBV11-2*01&quot;</td><td>&quot;TRBJ1-1*01&quot;</td><td>&quot;CASSSGGWGGGTEAFF&quot;</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td><td>5.45</td><td>6.38</td><td>4.37</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (21, 35)\n",
       "┌────────────┬─────────┬────────────┬───────────┬───┬──────────┬──────────┬────────────┬───────────┐\n",
       "│ job_name   ┆ cognate ┆ peptide    ┆ mhc_class ┆ … ┆ organism ┆ cdr_rmsd ┆ cdr_rmsd_a ┆ cdr_rmsd_ │\n",
       "│ ---        ┆ ---     ┆ ---        ┆ ---       ┆   ┆ ---      ┆ ---      ┆ f2_full    ┆ af2_trim  │\n",
       "│ str        ┆ bool    ┆ str        ┆ str       ┆   ┆ str      ┆ f64      ┆ ---        ┆ ---       │\n",
       "│            ┆         ┆            ┆           ┆   ┆          ┆          ┆ f64        ┆ f64       │\n",
       "╞════════════╪═════════╪════════════╪═══════════╪═══╪══════════╪══════════╪════════════╪═══════════╡\n",
       "│ 00d671d629 ┆ true    ┆ SPSYAYHQF  ┆ I         ┆ … ┆ mouse    ┆ 4.2      ┆ 11.69      ┆ 26.58     │\n",
       "│ eff4094fee ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 1be3c7f4da ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 66588fb6f4 ┆ true    ┆ RLQSLQTYV  ┆ I         ┆ … ┆ human    ┆ 8.07     ┆ 22.91      ┆ 26.8      │\n",
       "│ 1c42796a59 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ e21f801dcc ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ c874eda106 ┆ true    ┆ SLLMWITQV  ┆ I         ┆ … ┆ human    ┆ 5.67     ┆ 23.64      ┆ 20.27     │\n",
       "│ 179b27d613 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 48b7557431 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 1e2f70e91c ┆ true    ┆ FEDLRVLSF  ┆ I         ┆ … ┆ human    ┆ 2.49     ┆ 5.28       ┆ 3.14      │\n",
       "│ 8b2d204ab7 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ d420da955a ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 7f4b0f1dd8 ┆ true    ┆ ASNENMETM  ┆ I         ┆ … ┆ mouse    ┆ 8.6      ┆ 25.94      ┆ 16.93     │\n",
       "│ f99a68601d ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 93e79d9738 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆ …       ┆ …          ┆ …         ┆ … ┆ …        ┆ …        ┆ …          ┆ …         │\n",
       "│ cce720e42d ┆ true    ┆ SYYGPKTSPV ┆ II        ┆ … ┆ mouse    ┆ 3.83     ┆ 3.89       ┆ 4.58      │\n",
       "│ 01d1e88203 ┆         ┆ Q          ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 18af338c7c ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ da51f41cc3 ┆ true    ┆ APFSEQEQPV ┆ II        ┆ … ┆ human    ┆ 3.9      ┆ 2.81       ┆ 3.33      │\n",
       "│ 6cf890da26 ┆         ┆ L          ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 880c6f14fe ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ b05a70ac15 ┆ true    ┆ APMPMPELPY ┆ II        ┆ … ┆ human    ┆ 3.2      ┆ 4.52       ┆ 4.53      │\n",
       "│ dcec9c3125 ┆         ┆ P          ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ e3649535e8 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 08a74d65f9 ┆ true    ┆ VVQSELPYPE ┆ II        ┆ … ┆ human    ┆ 3.03     ┆ 20.29      ┆ 5.43      │\n",
       "│ e5f41a5ec6 ┆         ┆ G          ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ 7c0accfde9 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ a002927670 ┆ true    ┆ APFSEQEQPV ┆ II        ┆ … ┆ human    ┆ 5.45     ┆ 6.38       ┆ 4.37      │\n",
       "│ 058c8f7dc0 ┆         ┆ L          ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ ebffd01939 ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "│ …          ┆         ┆            ┆           ┆   ┆          ┆          ┆            ┆           │\n",
       "└────────────┴─────────┴────────────┴───────────┴───┴──────────┴──────────┴────────────┴───────────┘"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "af2_cutoff = pl.lit(datetime(2018, 6, 1, tzinfo=timezone.utc))\n",
    "\n",
    "phil_pdb.filter(pl.col(\"pdb_date\") > af2_cutoff)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "phil_pdb.write_csv(\n",
    "    \"elife_replicate/phil_pdb.csv\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|██████████| 22/22 [00:01<00:00, 18.76it/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (11, 26)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>job_name</th><th>cognate</th><th>peptide</th><th>mhc_class</th><th>mhc_1_chain</th><th>mhc_1_species</th><th>mhc_1_name</th><th>mhc_1_seq</th><th>mhc_2_chain</th><th>mhc_2_species</th><th>mhc_2_name</th><th>mhc_2_seq</th><th>tcr_1_chain</th><th>tcr_1_species</th><th>tcr_1_seq</th><th>tcr_2_chain</th><th>tcr_2_species</th><th>tcr_2_seq</th><th>pdb</th><th>pdb_date</th><th>antigen_chain</th><th>mhc_chain1</th><th>mhc_chain2</th><th>Achain</th><th>Bchain</th><th>organism</th></tr><tr><td>str</td><td>bool</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>datetime[μs, UTC]</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;39bf4e6e65f6b042d9340128ecd790…</td><td>true</td><td>&quot;GQVELGGGNAVEVCKG&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*03:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYSHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*03:02&quot;</td><td>&quot;GSGGSRDSPEDFVYQFKGMCYFTNGTERVR…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVTQTPRYLIKTRGQQVTLSCSPISGHRSV…</td><td>&quot;8vd0&quot;</td><td>2023-12-14 00:00:00 UTC</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;C&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;3dad0fd6a6e0ce25e4106fda13a57a…</td><td>true</td><td>&quot;VVGAVGVGK&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*11:100&quot;</td><td>&quot;GSHSMRYFYTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;QQKVQQSPESLIVPEGGMASLNCTSSDRNV…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;EAAVTQSPRNKVAVTGGKVTLSCNQTNNHN…</td><td>&quot;8i5c&quot;</td><td>2023-01-24 00:00:00 UTC</td><td>&quot;M&quot;</td><td>&quot;K&quot;</td><td>&quot;L&quot;</td><td>&quot;N&quot;</td><td>&quot;O&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;55461d93e2317bf4eecc407caf3910…</td><td>true</td><td>&quot;VVGAVGVGK&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*11:100&quot;</td><td>&quot;GSHSMRYFYTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;SWALSVHEGESVTVNCSYKTSITALQWYRQ…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;GVIQTPRHKVTGKGQEATLWCEPISGHSAV…</td><td>&quot;8i5d&quot;</td><td>2023-01-25 00:00:00 UTC</td><td>&quot;P&quot;</td><td>&quot;H&quot;</td><td>&quot;L&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;7ed224bf27775709b488a69d90cbe5…</td><td>true</td><td>&quot;PKYVKQNTLKLAR&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DRA*01:01&quot;</td><td>&quot;MIKEEHVIIQAEFYLNPDQSGEFMFDFDGD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DRB1*01:01&quot;</td><td>&quot;MGDTRPRFLWQLKFECHFFNGTERVRLLER…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MAQSVTQLGSHVSVSEGALVLLRCNYSSSV…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;MADVTQTPRNRITKTGKRIMLECSQTKGHD…</td><td>&quot;8pjg&quot;</td><td>2023-06-23 00:00:00 UTC</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;df790dd4eb9a5164bc3171d1ce5f70…</td><td>true</td><td>&quot;GQVELGGGTPIESCQ&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*03:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYSHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*03:02&quot;</td><td>&quot;RDSPEDFVYQFKGMCYFTNGTERVRLVTRY…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVTQTPRYLIKTRGQQVTLSCSPISGHRSV…</td><td>&quot;8vd2&quot;</td><td>2023-12-14 00:00:00 UTC</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;6899d34d074bdb896dbde93243ae73…</td><td>true</td><td>&quot;NYNYLYRLF&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*24:02&quot;</td><td>&quot;GSHSMRYFSTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;EVEQDPGPFNVPEGATVAFNCTYSNSASQS…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;AGVTQTPKFQVLKTGQSMTLQCAQDMNHNS…</td><td>&quot;8ye4&quot;</td><td>2024-02-21 00:00:00 UTC</td><td>&quot;E&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;G&quot;</td><td>&quot;H&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;98daa70e8e2597cfecfacfff18353f…</td><td>true</td><td>&quot;GQVELGGGPGAESCQ&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*03:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYSHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*03:02&quot;</td><td>&quot;RDSPEDFVYQFKGMCYFTNGTERVRLVTRY…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVTQTPRYLIKTRGQQVTLSCSPISGHRSV…</td><td>&quot;8vcx&quot;</td><td>2023-12-14 00:00:00 UTC</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;404f58ae3d26a08ecc8bd707789d01…</td><td>true</td><td>&quot;VVGAVGVGK&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*11:01&quot;</td><td>&quot;GSHSMRYFYTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;QQKVQQSPESLIVPEGGMASLNCTSSDRNV…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;KIIQKPKYLVAVTGSEKILICEQYLGHNAM…</td><td>&quot;8wul&quot;</td><td>2023-10-20 00:00:00 UTC</td><td>&quot;Q&quot;</td><td>&quot;O&quot;</td><td>&quot;P&quot;</td><td>&quot;E&quot;</td><td>&quot;F&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;13f7a9410810562e44920fa0413159…</td><td>true</td><td>&quot;VVGAVGVGK&quot;</td><td>&quot;I&quot;</td><td>&quot;heavy&quot;</td><td>&quot;human&quot;</td><td>&quot;A*11:01&quot;</td><td>&quot;GSHSMRYFYTSVSRPGRGEPRFIAVGYVDD…</td><td>&quot;light&quot;</td><td>&quot;human&quot;</td><td>&quot;B2M&quot;</td><td>&quot;IQRTPKIQVYSRHPAENGKSNFLNCYVSGF…</td><td>&quot;alpha&quot;</td><td>&quot;mouse&quot;</td><td>&quot;QQKVQQSPESLIVPEGGMASLNCTSSDRNV…</td><td>&quot;beta&quot;</td><td>&quot;mouse&quot;</td><td>&quot;KIIQKPKYLVAVTGSEKILICEQYLGHNAM…</td><td>&quot;8wte&quot;</td><td>2023-10-18 00:00:00 UTC</td><td>&quot;J&quot;</td><td>&quot;H&quot;</td><td>&quot;I&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;human&quot;</td></tr><tr><td>&quot;e6da04b0f02efa07a5e1795130328a…</td><td>true</td><td>&quot;GQVELGGGSSPETCI&quot;</td><td>&quot;II&quot;</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;DQA1*03:01&quot;</td><td>&quot;EDIVADHVASYGVNLYQSYGPSGQYSHEFD…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;DQB1*03:02&quot;</td><td>&quot;RDSPEDFVYQFKGMCYFTNGTERVRLVTRY…</td><td>&quot;alpha&quot;</td><td>&quot;human&quot;</td><td>&quot;MKTTQPPSMDCAEGRAANLPCNHSTISGNE…</td><td>&quot;beta&quot;</td><td>&quot;human&quot;</td><td>&quot;GVTQTPRYLIKTRGQQVTLSCSPISGHRSV…</td><td>&quot;8vcy&quot;</td><td>2023-12-14 00:00:00 UTC</td><td>&quot;C&quot;</td><td>&quot;A&quot;</td><td>&quot;B&quot;</td><td>&quot;D&quot;</td><td>&quot;E&quot;</td><td>&quot;human&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (11, 26)\n",
       "┌───────────────┬─────────┬──────────────┬───────────┬───┬────────────┬────────┬────────┬──────────┐\n",
       "│ job_name      ┆ cognate ┆ peptide      ┆ mhc_class ┆ … ┆ mhc_chain2 ┆ Achain ┆ Bchain ┆ organism │\n",
       "│ ---           ┆ ---     ┆ ---          ┆ ---       ┆   ┆ ---        ┆ ---    ┆ ---    ┆ ---      │\n",
       "│ str           ┆ bool    ┆ str          ┆ str       ┆   ┆ str        ┆ str    ┆ str    ┆ str      │\n",
       "╞═══════════════╪═════════╪══════════════╪═══════════╪═══╪════════════╪════════╪════════╪══════════╡\n",
       "│ 39bf4e6e65f6b ┆ true    ┆ GQVELGGGNAVE ┆ II        ┆ … ┆ C          ┆ D      ┆ E      ┆ human    │\n",
       "│ 042d9340128ec ┆         ┆ VCKG         ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ d790…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 3dad0fd6a6e0c ┆ true    ┆ VVGAVGVGK    ┆ I         ┆ … ┆ L          ┆ N      ┆ O      ┆ human    │\n",
       "│ e25e4106fda13 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ a57a…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 55461d93e2317 ┆ true    ┆ VVGAVGVGK    ┆ I         ┆ … ┆ L          ┆ A      ┆ B      ┆ human    │\n",
       "│ bf4eecc407caf ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 3910…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 7ed224bf27775 ┆ true    ┆ PKYVKQNTLKLA ┆ II        ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ 709b488a69d90 ┆         ┆ R            ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ cbe5…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ df790dd4eb9a5 ┆ true    ┆ GQVELGGGTPIE ┆ II        ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ 164bc3171d1ce ┆         ┆ SCQ          ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 5f70…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ …             ┆ …       ┆ …            ┆ …         ┆ … ┆ …          ┆ …      ┆ …      ┆ …        │\n",
       "│ 6899d34d074bd ┆ true    ┆ NYNYLYRLF    ┆ I         ┆ … ┆ B          ┆ G      ┆ H      ┆ human    │\n",
       "│ b896dbde93243 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ ae73…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 98daa70e8e259 ┆ true    ┆ GQVELGGGPGAE ┆ II        ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ 7cfecfacfff18 ┆         ┆ SCQ          ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 353f…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 404f58ae3d26a ┆ true    ┆ VVGAVGVGK    ┆ I         ┆ … ┆ P          ┆ E      ┆ F      ┆ human    │\n",
       "│ 08ecc8bd70778 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 9d01…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 13f7a94108105 ┆ true    ┆ VVGAVGVGK    ┆ I         ┆ … ┆ I          ┆ A      ┆ B      ┆ human    │\n",
       "│ 62e44920fa041 ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 3159…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ e6da04b0f02ef ┆ true    ┆ GQVELGGGSSPE ┆ II        ┆ … ┆ B          ┆ D      ┆ E      ┆ human    │\n",
       "│ a07a5e1795130 ┆         ┆ TCI          ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "│ 328a…         ┆         ┆              ┆           ┆   ┆            ┆        ┆        ┆          │\n",
       "└───────────────┴─────────┴──────────────┴───────────┴───┴────────────┴────────┴────────┴──────────┘"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tmp = serial_apply(\n",
    "    post_cutoff_pdb,\n",
    "    get_pdb_date,\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "af3-analyzer",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
